File: src/Gostio.Model/Validation/ImageRules.cs
using Gostio.Model.Exceptions;
using Gostio.Model.Requests;

namespace Gostio.Model.Validation;

public static class ImageRules
{
    public const int MaximumBytes = 4 * 1024 * 1024;

    public const string Jpeg = "image/jpeg";

    public const string Png = "image/png";

    public const string Webp = "image/webp";

    // What a client sends when it did not look at the file, so it is the
    // absence of a claim rather than a wrong one.
    public const string Unknown = "application/octet-stream";

    public static IReadOnlyList<string> Allowed { get; } = [Jpeg, Png, Webp];

    public static string? Detect(ReadOnlySpan<byte> content)
    {
        if (content.Length >= 3 && content[0] == 0xFF && content[1] == 0xD8 && content[2] == 0xFF)
        {
            return Jpeg;
        }

        if (content.Length >= 8 && content[..8].SequenceEqual(PngSignature))
        {
            return Png;
        }

        if (content.Length >= 12
            && content[..4].SequenceEqual("RIFF"u8)
            && content[8..12].SequenceEqual("WEBP"u8))
        {
            return Webp;
        }

        return null;
    }

    public static string RequireImage(ImageUpload upload, string field)
    {
        if (upload.Content.Length == 0)
        {
            throw new ValidationException(field, "Choose an image to upload.");
        }

        if (upload.Content.Length > MaximumBytes)
        {
            throw new ValidationException(
                field, $"An image is at most {MaximumBytes / (1024 * 1024)} MB.");
        }

        var detected = Detect(upload.Content)
            ?? throw new ValidationException(
                field, $"An image has to be one of {string.Join(", ", Allowed)}.");

        // The claim is checked and then dropped: what reaches the column is
        // what the bytes proved, so a stored type holds on the way back out.
        var claimed = Claimed(upload.ContentType);

        if (claimed is not null
            && !string.Equals(claimed, detected, StringComparison.OrdinalIgnoreCase))
        {
            throw new ValidationException(
                field, $"This file was sent as {claimed} and its bytes say {detected}.");
        }

        return detected;
    }

    private static string? Claimed(string? contentType)
    {
        var named = contentType?.Split(';')[0].Trim();

        return string.IsNullOrEmpty(named)
            || string.Equals(named, Unknown, StringComparison.OrdinalIgnoreCase)
                ? null
                : named;
    }

    private static ReadOnlySpan<byte> PngSignature =>
        [0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A];
}

File: apps/gostio_desktop/lib/core/validation/image_rules.dart
import 'dart:typed_data';

// What the server accepts as an uploaded image, mirrored so a file it would
// refuse is refused here first and named the same way.
abstract final class ImageRules {
  static const int maximumBytes = 4 * 1024 * 1024;

  static const String jpeg = 'image/jpeg';
  static const String png = 'image/png';
  static const String webp = 'image/webp';

  static const String unknown = 'application/octet-stream';

  static const List<String> allowed = <String>[jpeg, png, webp];

  static const List<String> extensions = <String>['jpg', 'jpeg', 'png', 'webp'];

  static String? detect(Uint8List content) {
    if (_matches(content, 0, _jpegSignature)) {
      return jpeg;
    }

    if (_matches(content, 0, _pngSignature)) {
      return png;
    }

    return _matches(content, 0, _riffTag) && _matches(content, 8, _webpTag)
        ? webp
        : null;
  }

  static String? refuse(Uint8List content) {
    if (content.isEmpty) {
      return 'Choose an image to upload.';
    }

    if (content.length > maximumBytes) {
      return 'An image is at most ${maximumBytes ~/ (1024 * 1024)} MB.';
    }

    return detect(content) == null
        ? 'An image has to be one of ${allowed.join(', ')}.'
        : null;
  }

  static bool _matches(Uint8List content, int at, List<int> signature) {
    if (content.length < at + signature.length) {
      return false;
    }

    for (int index = 0; index < signature.length; index++) {
      if (content[at + index] != signature[index]) {
        return false;
      }
    }

    return true;
  }

  static const List<int> _jpegSignature = <int>[0xFF, 0xD8, 0xFF];
  static const List<int> _pngSignature = <int>[
    0x89,
    0x50,
    0x4E,
    0x47,
    0x0D,
    0x0A,
    0x1A,
    0x0A,
  ];
  static const List<int> _riffTag = <int>[0x52, 0x49, 0x46, 0x46];
  static const List<int> _webpTag = <int>[0x57, 0x45, 0x42, 0x50];
}
